test/algorithms/classifiers/test_vqc.py

```diff
@@ -12,390 +12,253 @@
+    """VQC Tests."""
+
+    # Set-up the quantum instances.
+    sv_quantum_instance = QuantumInstance(
+        Aer.get_backend("aer_simulator_statevector"),
+        seed_simulator=algorithm_globals.random_seed,
+        seed_transpiler=algorithm_globals.random_seed,
     )
-    def test_default_parameters(self, config):
-        """Test VQC instantiation with default parameters."""
+    qasm_quantum_instance = QuantumInstance(
+        Aer.get_backend("aer_simulator"),
+        shots=100,
+        seed_simulator=algorithm_globals.random_seed,
+        seed_transpiler=algorithm_globals.random_seed,
+    )
+    quantum_instances = [sv_quantum_instance, qasm_quantum_instance]
 
-        provide_num_qubits, provide_feature_map, provide_ansatz = config
-        num_inputs = 2
+    # Set-up the numbers of qubits.
+    num_qubits_list = [2, None]
 
-        num_qubits, feature_map, ansatz = None, None, None
+    # Set-up the feature maps.
+    num_features = 2
+    zz_feature_map = ZZFeatureMap(num_features)
+    feature_maps = [zz_feature_map, None]
 
-        if provide_num_qubits:
-            num_qubits = num_inputs
-        if provide_feature_map:
-            feature_map = ZZFeatureMap(num_inputs)
-        if provide_ansatz:
-            ansatz = RealAmplitudes(num_inputs, reps=1)
+    # Set-up the ansatzes
+    real_amplitudes = RealAmplitudes(num_features, reps=1)
+    ansatzes = [real_amplitudes, None]
 
-        classifier = VQC(
-            num_qubits=num_qubits,
-            feature_map=feature_map,
-            ansatz=ansatz,
-            quantum_instance=self.qasm_quantum_instance,
-        )
+    # Set-up the optimizers.
+    bfgs = L_BFGS_B(maxiter=5)
+    cobyla = COBYLA(maxiter=25)
+    optimizers = [bfgs, cobyla, None]
 
-        # construct data
-        num_samples = 5
-        # pylint: disable=invalid-name
-        X = algorithm_globals.random.random((num_samples, num_inputs))
-        y = 1.0 * (np.sum(X, axis=1) <= 1)
-        while len(np.unique(y)) == 1:
-            X = algorithm_globals.random.random((num_samples, num_inputs))
-            y = 1.0 * (np.sum(X, axis=1) <= 1)
-        y = np.array([y, 1 - y]).transpose()  # VQC requires one-hot encoded input
-
-        # fit to data
-        classifier.fit(X, y)
-
-        # score
-        score = classifier.score(X, y)
-        self.assertGreater(score, 0.5)
+    # Set-up the loss functions.
+    losses = ["squared_error", "absolute_error", "cross_entropy"]
 
-    @data(
-        # optimizer, quantum instance
-        ("cobyla", "statevector"),
-        ("cobyla", "qasm"),
-        ("bfgs", "statevector"),
-        ("bfgs", "qasm"),
-        (None, "statevector"),
-        (None, "qasm"),
+    # Set-up the datasets.
+    binary_x, binary_y = _generate_classification_data(6, 2, num_features)
+    binary_datasets = [(binary_x, _one_hot_encode(binary_y))]
+    multiclass_x, multiclass_y = _generate_classification_data(9, 3, num_features)
```

Comment:
Curious, why you choose these sizes, 6 and 9. Is there a reason for that? I don't say this wrong, I think, in the initial tests sizes were different.

Reply:
Ok, so partly this is due to OCD. I found that I had to increase the number of samples to 9 in the 3 class classification so it can score above `0.5`, rather than `1 / num_classes`. I'm not totally sure it's guaranteed to pass as I think there is still stochasticity in the tests, despite fixing the seed. So, then I made the 2 class classification 6 so that it was linearly proportional. I can probably reduce it.

Reply:
Usually more samples increases stability. But too many samples increases training time, so there's a trade off in the tests. If it works, fine with me.

Reply:
Is there any way we remove all the randomness from the optimization? Are there additional seeds we can set?

Reply:
There are two seeds in `QuantumInstance`, randomness in the data generation and `initial_point`. This is what comes to my mind. Seems like everything is fixed.

Reply:
The failing test does not have randomization in the code. So everything is fixed upfront. Fixing a seed in `setUp()` may or may not help.

Reply:
There is a seed there set up front as a class level variable is what I was referring to above (that was a copy of line in question) in moving it. Where it currently is anything later that uses the algorithms random generator will be subject to loading/execution order e.g. _generate_classification_data that uses it.  I saw comments around random behvaior, and maybe the discussion was over a specific test for which this comment does not specifically apply. I think it does in general though.

Reply:
There's another issue with the tests. Old tests (main branch) take roughly 1:30, new tests 5:10 on my laptop. There are more new tests here, but I suspect the datasets are not great and fitting them, especially with LBFGS, takes quite some time.

Reply:
I've hardcoded the data with some (hopefully consistently) working examples with two data points for each class. If we want to make it quicker, we could remove some of the other test variations, if they're not providing much extra value.

Reply:
Ok, I've decided to change 0.5 back to 1 / num_classes, as it was before. Given the unit test speed requirements, I think it's ok just to check that the score is somewhat reasonable given the minimal sample size. It's probably best to evaluate performance properly outside the unit tests. Let me know if you disagree though.

Reply:
I suggest to generate data somehow differently. The original dataset seems to be a bit noisy.
```python
from sklearn.datasets import make_classification
from sklearn.preprocessing import MinMaxScaler

BINARY_X, BINARY_Y = make_classification(n_samples=6, n_features=2, n_classes=2, n_redundant=0, n_clusters_per_class=1, class_sep=5.0, random_state=algorithm_globals.random_seed)
BINARY_X = MinMaxScaler().fit_transform(BINARY_X)

MULTICLASS_X, MULTICLASS_Y = make_classification(n_samples=9, n_features=2, n_classes=3, n_redundant=0, n_clusters_per_class=1, class_sep=5.0, random_state=algorithm_globals.random_seed)
MULTICLASS_X = MinMaxScaler().fit_transform(MULTICLASS_X)
```

You can increase `class_sep` to make classes more separable.

I replaced the original data generation lines with this snippet and in a few cases I get score of `0.88`, mostly it is `1.0`. Reducing the number of samples to 4 and 5 respectively adds more instability in the test results. In the warm start test I got a score of 0.5. So, I think 6 and 9 are reasonable. With these datasets you may test a score against `0.5`, not `1 / num_classes`.

Also, I suggest to remove `None` from a list of optimizers and have a non-parameterized dedicated unit test where we don't set optimizer. The reason is that if optimizer is not set then a default one is created with no `maxiter` set. And this optimizer may cause some tests to run for a long time. Even without `None` the run time is not close to the original time, but somewhat lesser, around 3:20 on my laptop. Reducing `maxiter` in `LBFGS` and `COBYLA` to 3 and 15 respectively is still reasonable. I observe quite a number of scores of `0.77` and still a lot of `1.0` and the run time is 2:10. I don't pretend that my numbers are statistically correct. In fact, there are more tests now than in the original code, so some performance degradation seems inevitable.

You may replace `_one_hot_encode` with `OneHotEncoder(sparse=False)` just for consistency, since I have `MixMaxScaler` in my snippet, but this is a minor.

Let me know if you have questions.

Reply:
Thanks, Anton. That helps a lot :).